src/LibraryManager/Utilities/ParallelUtility.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Web.LibraryManager.Utilities
+{
+    internal static class ParallelUtility
+    {
+        /// <summary>
+        /// Executes an ation over each element of a collection, up to the specified degree of parallelism.
+        /// </summary>
+        /// <param name="action">The action to be applied to each element</param>
+        /// <param name="degreeOfParallelism">How many tasks to run in parallel</param>
+        /// <param name="items">The items to </param>
+        public static Task ForEachAsync<T>(Func<T, Task> action, int degreeOfParallelism, IEnumerable<T> items)
+        {
+            return ForEachAsync<T>(action, degreeOfParallelism, items, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Executes an ation over each element of a collection, up to the specified degree of parallelism.
+        /// </summary>
+        /// <param name="action">The action to be applied to each element</param>
+        /// <param name="degreeOfParallelism">How many tasks to run in parallel</param>
+        /// <param name="items">The items to </param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        public static async Task ForEachAsync<T>(Func<T, Task> action, int degreeOfParallelism, IEnumerable<T> items, CancellationToken cancellationToken)
+        {
+            action = action ?? throw new ArgumentNullException(nameof(action));
+            items = items ?? throw new ArgumentNullException(nameof(items));
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var semaphore = new SemaphoreSlim(degreeOfParallelism, degreeOfParallelism);
+            var allTasks = new List<Task>();
+
+            foreach (T item in items)
+            {
+                await semaphore.WaitAsync();
+                cancellationToken.ThrowIfCancellationRequested();
+
+                Task task = DoAction(action, item, semaphore);
```

Comment:
Nit: I personally would've preferred to see Release call in this method (since WaitAsync is in this method) rather than in the helper method so that it's clear we are properly releasing. Or else I would've renamed your helper into something like DoActionAndRelease, again, so it's clear we are always releasing. So my suggestion would to either get rid of helper or rename it to make it clearer that it releases the semaphore.

Reply:
I can't release here.  There isn't a point in this method that I know the task has completed until after the `await Task.WhenAll()` below.  It's an implicit handshake between this method (using `semaphore.WaitAsync()`) to know when it can queue another job, and each job itself notifying when it's complete (`sempaphore.Release()`).  And I don't want to await each task as it's queued to be able to `Release` here without losing the parallel aspect (or, I end up chunking jobs, but that could mean that 9 files of <1KB are in the same chunk as a 1MB resource file, so I still lose parallel execution while waiting for the one big file; this is a realistic case with some larger libraries).  Also, if I call `WaitAsync()` in the `DoAction` helper, I'll end up queueing all the tasks immediately, which would add pressure to the scheduler when they'd all just get stuck waiting anyway.

The rename to `DoActionAndRelease` sounds reasonable.

Reply:
You are right, we can't move Release up. Renaming the method sounds great.